### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by jeberl**
Summary
---------

Adding TypeScript to our webapp. Benefit of typescript is that it doens't have to be used universally. A Typescript file (.tsx) can interact with .jsx files and visa versa. This means that we can transfer over to typescript piecemeal

Converted one file to TypeScript as a proof of concept. ActionRuleFormColumns.tsx 

In doing so, needed to upgrade to a newer version of prettier which caused a few changes elsewhere.

Also needed to switch to the typescript parser for ESLint which required some other changes elsewhere.

Test Plan
---------

- Show that typescript will work natively in devcontainer and will show typing errors
- Show that typescript files will build
- Show that prettier will auto-format on save for .jsx files
- Show that prettier will auto-format on save for .tsx files also
- Check all pages changed in UI and ensure still work the same

### CODE REVIEW

File: hasher-matcher-actioner/webapp/.eslintrc.json [hunk 1]
```diff
@@ -1,24 +1,35 @@
...
+      "jsx": true,
+      "tsx": true
     },
     "ecmaVersion": 12,
     "sourceType": "module"
   },
-  "plugins": ["react", "prettier"],
+  "plugins": ["react", "prettier", "@typescript-eslint"],
   "rules": {
+    "no-use-before-define": "off",
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/.eslintrc.json:27]: The eslint error `no-use-before-define` breaks when using the `typescript-eslint` parser so we override it with `typescript-eslint`'s own version of the error. This is direct from the `typescript-eslint` documentation

File: hasher-matcher-actioner/webapp/package.json [hunk 2]
```diff
@@ -57,6 +61,7 @@
     "eslint-plugin-prettier": "^3.3.1",
     "eslint-plugin-react": "^7.22.0",
     "eslint-plugin-react-hooks": "^4.2.0",
-    "prettier": "^2.2.1"
+    "prettier": "2.3.1",
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/package.json:64]: typescript requires a higher version of prettier

File: hasher-matcher-actioner/webapp/src/components/HolidaysDatasetInformationBlock.jsx [hunk 3]
```diff
@@ -21,7 +21,7 @@ export function HolidaysDatasetInformationBlock({samplePGExists, refresh}) {
       `This group will become disabled on "Sync" as it does not have TE
       counterpart. Simply delete and create again if you wish to keep match
       sample data after syncing`,
-    ).then(() => {});
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/components/HolidaysDatasetInformationBlock.jsx:24]: A eslint requested change. Tested this by deleting and recreating the holiday dataset in the UI

File: hasher-matcher-actioner/webapp/src/components/HolidaysDatasetInformationBlock.jsx [hunk 4]
```diff
@@ -72,5 +74,5 @@ HolidaysDatasetInformationBlock.propTypes = {
 
 HolidaysDatasetInformationBlock.defaultProps = {
   samplePGExists: false,
-  refresh: () => {},
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/components/HolidaysDatasetInformationBlock.jsx:75]: eslint doenst like do-nothing functions so made nullable

File: hasher-matcher-actioner/webapp/src/components/OpinionTableCell.jsx [hunk 5]
```diff
@@ -55,16 +55,14 @@ export default function OpinionTableCell({
...
-              ? () => {}
-              : () => {
-                  setNewPendingOpinionChange(
-                    PENDING_OPINION_CHANGE.MARK_FALSE_POSITIVE,
-                  );
-                  setShowModal(true);
-                }
-          }>
+          onClick={() => {
+            if (!updatePending) {
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/components/OpinionTableCell.jsx:59]: Dito on do nothing functions

File: hasher-matcher-actioner/webapp/src/components/OpinionTableCell.jsx [hunk 6]
```diff
@@ -80,16 +78,14 @@ export default function OpinionTableCell({
             updatePending &&
             pendingOpinionChange === PENDING_OPINION_CHANGE.MARK_TRUE_POSITIVE
           }
-          onClick={
-            updatePending
-              ? () => {}
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/components/OpinionTableCell.jsx:85]: .

File: hasher-matcher-actioner/webapp/src/components/settings/ActionPerformer/ActionPerformerRows.jsx [hunk 7]
```diff
@@ -20,14 +20,10 @@ export default function ActionPerformerRows({
   canNotDeleteOrUpdateName,
 }) {
   const [editing, setEditing] = useState(edit);
-  const [
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/components/settings/ActionPerformer/ActionPerformerRows.jsx:23]: formatting changes

File: hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx [hunk 8]
```diff
@@ -3,14 +3,36 @@
  */
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:4]: This is the one file that is now TypeScript

File: hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx [hunk 9]
```diff
@@ -101,7 +123,7 @@ export default function ActionRuleFormColumns({
       <td>
         <Form.Label>
           Name
-          <span hidden={!showErrors || name}> (required)</span>
+          <span hidden={!showErrors || !!name}> (required)</span>
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:126]: `!showErrors || name` returns a string if `showErrors` is true

`!showErrors || !!name` always returns a bool

File: hasher-matcher-actioner/webapp/src/pages/Dashboard.jsx [hunk 10]
```diff
@@ -69,42 +69,31 @@ function toUFlotFormat(graphData) {
   return [timestamps, values];
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/pages/Dashboard.jsx:69]: These changes were just ordering. Dont use a function before you define it. Tested by looking at Dashboard

File: hasher-matcher-actioner/webapp/src/pages/Matches.jsx [hunk 11]
```diff
@@ -29,71 +29,6 @@ import {useQuery} from '../utils/QueryParams';
 import FullWidthLocalScrollingLeftAlignedLayout from './layouts/FullWidthLocalScrollingLeftAlignedLayout';
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/pages/Matches.jsx:29]: Same. Just moving functions around

File: hasher-matcher-actioner/webapp/src/pages/settings/SignalSettingsTab.jsx [hunk 12]
```diff
@@ -10,11 +10,48 @@ import ListGroup from 'react-bootstrap/ListGroup';
 import Button from 'react-bootstrap/Button';
```
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/pages/settings/SignalSettingsTab.jsx:10]: More moving around

**Review by eriksendc — APPROVED** [review]
Looks good to me and I approve. But check with @BarrettOlson to see if he wants to review before you consider this done.

File: hasher-matcher-actioner/webapp/.eslintrc.json (hunk 1 above)
Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/.eslintrc.json:27]: nit: link to documentation/if comments allowed provide the context here for future readers.

File: hasher-matcher-actioner/webapp/package.json (hunk 2 above)
Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/package.json:64]: nit: might was well leave the `^`

File: hasher-matcher-actioner/webapp/src/components/HolidaysDatasetInformationBlock.jsx (hunk 3 above)
Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/src/components/HolidaysDatasetInformationBlock.jsx:24]: hmm, I think not waiting on the future at all is a slight change in logic but the only case I could maybe see that mattering is if they click off the page immediately. Probably fine

File: hasher-matcher-actioner/webapp/.eslintrc.json [hunk 13]
```diff
@@ -1,24 +1,35 @@
...
+    "@typescript-eslint/no-use-before-define": ["error"],
     "react/jsx-filename-extension": [
       "warn",
       {
-        "extensions": [".js", ".jsx"]
+        "extensions": [".js", ".jsx", ".tsx"]
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/.eslintrc.json:32]: to keep the pattern consistent/incase we end up having .ts files can you add them as well?

File: hasher-matcher-actioner/webapp/.prettierrc.json [hunk 14]
```diff
@@ -7,7 +7,7 @@
   "endOfLine": "auto",
   "overrides": [
     {
-      "files": ["*.js", "*.jsx"]
+      "files": ["*.js", "*.jsx", "*.tsx"]
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/.prettierrc.json:10]: .

File: hasher-matcher-actioner/webapp/README.md [hunk 15]
```diff
@@ -57,7 +57,7 @@ Please run the following checks/formating before submitting a PR.
 
 ```
 # linter
-npx eslint src --ext .js,.jsx --fix
+npx eslint src --ext .js,.jsx,.tsx --fix
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/README.md:60]: .

File: hasher-matcher-actioner/webapp/package.json [hunk 16]
```diff
@@ -14,13 +14,16 @@
     "clipboard-copy": "^4.0.1",
     "date-fns": "^2.21.0",
     "dompurify": "^2.2.8",
+    "eslint": "^7.29.0",
     "node-sass": "^5.0.0",
+    "prettier": "^2.3.1",
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/package.json:19]: nit: I don't know if these needed to added beyond devDependencies do you have errors without them?

File: hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx [hunk 17]
```diff
@@ -3,14 +3,36 @@
...
-import {PropTypes} from 'prop-types';
+import PropTypes from 'prop-types';
 import Form from 'react-bootstrap/Form';
 import Row from 'react-bootstrap/Row';
 import Col from 'react-bootstrap/Col';
 import Button from 'react-bootstrap/Button';
 
 export const classificationTypeTBD = 'TBD';
 
+type Action = {
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:14]: This is a larger concern probably beyond the scope of this PR but I would strongly prefer things like `Action` and `Classification` actually correspond to something on the backend in python. 

Right now these models makes sense for the UI but don't map back as well. The more the UI interface and the actually backing structure diverge the more difficult to develop on and imo less intuitive the system will become.

Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:14]: At a minimum can you add some docstrings to these types?

File: hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx [hunk 18]
```diff
@@ -3,14 +3,36 @@
...
+  id: string;
+  name: string;
+};
+
+type Classification = {
+  classificationType: string;
+  equals: boolean;
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:21]: is it confusing what this fields is saying, equal_to vs not_equal_to check when a match is found?

File: hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx [hunk 19]
```diff
@@ -60,7 +82,7 @@ export default function ActionRuleFormColumns({
         <Form.Control
           as="select"
           required
-          value={classification.equals}
+          value={classification.equals ? 'true' : 'false'}
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:85]: I think 'Equal' vs 'Not Equal' here might help

**Review by BarrettOlson — APPROVED** [review]
Nice! nits and such inline but nothing really about the typescript change itself. Bit of a shame this messes with the blame as much as it does but given we are adding typing needing `no-use-before-define` makes sense.

File: hasher-matcher-actioner/webapp/package.json (hunk 16 above)
Comment (jeberl) [reply hasher-matcher-actioner/webapp/package.json:19]: They were automatically generated. Can try without

File: hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx (hunk 17 above)
Comment (jeberl) [reply hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:14]: This doesnt map 1:1 with anything in python. Its effectively a pointer an Action. The Action Settings tab would have the 1:1 mapping with `WebhookActionPerformer` in python

File: hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx (hunk 18 above)
Comment (jeberl) [reply hasher-matcher-actioner/webapp/src/components/settings/ActionRuleFormColumns.tsx:21]: Gonna rename the class and the field

File: hasher-matcher-actioner/webapp/src/react-app-env.d.ts [hunk 20]
```diff
@@ -0,0 +1 @@
+/// <reference types="react-scripts" />
```
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/src/react-app-env.d.ts:1]: what does this do?

**Pull request merged**